### PR TITLE
feat: bump gwapi to v0.6.0

### PR DESCRIPTION
**Pull request by Xunzhuo**
Fixes: https://github.com/envoyproxy/gateway/issues/839
Fixes: https://github.com/envoyproxy/gateway/issues/716

Closes: https://github.com/envoyproxy/gateway/pull/828

Signed-off-by: bitliu <bitliu@tencent.com>

### CODE REVIEW

File: internal/provider/kubernetes/controller.go [hunk 1]
```diff
@@ -495,6 +495,14 @@ func (r *gatewayAPIReconciler) statusUpdateForGateway(gtw *gwapiv1b1.Gateway, sv
 			}
 			gCopy := g.DeepCopy()
 			gCopy.Status.Conditions = status.MergeConditions(gCopy.Status.Conditions, gtw.Status.Conditions...)
+			for index := range gCopy.Status.Conditions {
```
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:498]: thanks for fixing the issue @Xunzhuo !
I have two questions
* why are we updating all the conditions here ?
isn't this already done correctly in https://github.com/envoyproxy/gateway/blob/085e6fa5a56066472a52e8c07bc1aefa3ffca5a2/internal/status/conditions.go#L116
* shouldn't this logic be living in https://github.com/envoyproxy/gateway/blob/main/internal/status/conditions.go
cc @skriss  who was also looking into this

**Comment by arkodg** [comment]
looks like the core issue is interpretation of mapping of Observed Generation to Status Conditions
The new GAPI tests assert that the new Observed Generation value be asserted on all Status Conditions (https://github.com/kubernetes-sigs/gateway-api/pull/1586) but Envoy Gateway does not update the status conditions for all older conditions if the core values have not changed https://github.com/envoyproxy/gateway/blob/085e6fa5a56066472a52e8c07bc1aefa3ffca5a2/internal/status/conditions.go#L127 . Imho the latter approach seems more logical. Is there any spec governing this mapping @youngnick

**Comment by sunjayBhatia** [comment]
> looks like the core issue is interpretation of mapping of Observed Generation to Status Conditions The new GAPI tests assert that the new Observed Generation value be asserted on all Status Conditions ([kubernetes-sigs/gateway-api#1586](https://github.com/kubernetes-sigs/gateway-api/pull/1586)) but Envoy Gateway does not update the status conditions for all older conditions if the core fields (such as `Status`, `Reason` and `Message`) have not changed
> 
> https://github.com/envoyproxy/gateway/blob/085e6fa5a56066472a52e8c07bc1aefa3ffca5a2/internal/status/conditions.go#L127
> 
> and
> https://github.com/envoyproxy/gateway/blob/085e6fa5a56066472a52e8c07bc1aefa3ffca5a2/internal/gatewayapi/contexts.go#L484
> 
> . Imho the latter approach seems more logical. Is there any spec governing this mapping @youngnick @skriss

from the Condition spec: https://github.com/kubernetes/apimachinery/blob/6c409361e35e40e38c4056ba0b86647d4244c047/pkg/apis/meta/v1/types.go#L1480-L1485

**Comment by arkodg** [comment]
thanks for sharing that @sunjayBhatia ! , the spec makes it clear that the condition is considered out of date if the observed generation is older
hey @Xunzhuo if you can enhance these snippets (https://github.com/envoyproxy/gateway/blob/085e6fa5a56066472a52e8c07bc1aefa3ffca5a2/internal/status/conditions.go#L127 and https://github.com/envoyproxy/gateway/blob/085e6fa5a56066472a52e8c07bc1aefa3ffca5a2/internal/gatewayapi/contexts.go#L484) with an extra check to consider a condition unchanged or changed if the observed generation is also unchanged or changed respectively, should fix the issue, thanks !

File: internal/provider/kubernetes/controller.go (hunk 1 above)
Comment (arkodg) [reply internal/provider/kubernetes/controller.go:498]: please ignore this in favor of https://github.com/envoyproxy/gateway/pull/853#issuecomment-1371423935

**Comment by Xunzhuo** [comment]
> even if the type/reason/status haven't changed, the observed generation should be updated to make clear the status condition applies to the latest version of the resource
From this comment, do we still need to add an extra check to consider a condition unchanged or changed if the observed generation is also unchanged or changed respectively ?

**Comment by Xunzhuo** [comment]
> thanks for sharing that @sunjayBhatia ! , the spec makes it clear that the condition is considered out of date if the observed generation is older hey @Xunzhuo if you can enhance these snippets (
> 
> https://github.com/envoyproxy/gateway/blob/085e6fa5a56066472a52e8c07bc1aefa3ffca5a2/internal/status/conditions.go#L127
> 
> and
> https://github.com/envoyproxy/gateway/blob/085e6fa5a56066472a52e8c07bc1aefa3ffca5a2/internal/gatewayapi/contexts.go#L484
> 
> ) with an extra check to consider a condition unchanged or changed if the observed generation is also unchanged or changed respectively, should fix the issue, thanks !

Done.

File: internal/gatewayapi/contexts.go [hunk 2]
```diff
@@ -494,7 +494,7 @@ func (r *RouteParentContext) SetCondition(route RouteContext, conditionType v1be
 	idx := -1
 	routeStatus := route.GetRouteStatus()
 	for i, existing := range routeStatus.Parents[r.routeParentStatusIdx].Conditions {
-		if existing.Type == cond.Type {
+		if existing.Type == cond.Type && existing.ObservedGeneration == cond.ObservedGeneration {
```
Comment (arkodg) [inline internal/gatewayapi/contexts.go:497]: lines 499-501 need to be changed to
```
if existing.Status == cond.Status &&
 				existing.Reason == cond.Reason &&
				existing.Message == cond.Message &&
				existing.ObservedGeneration == cond.ObservedGeneration {
```

File: internal/provider/kubernetes/controller.go [hunk 3]
```diff
@@ -495,6 +495,14 @@ func (r *gatewayAPIReconciler) statusUpdateForGateway(gtw *gwapiv1b1.Gateway, sv
...
+				gCopy.Status.Conditions[index].ObservedGeneration = gtw.Generation
```
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:499]: this loop needs to be removed, this should happen automatically

File: internal/provider/kubernetes/controller.go [hunk 4]
```diff
@@ -894,6 +902,15 @@ func (r *gatewayAPIReconciler) subscribeAndUpdateStatus(ctx context.Context) {
 						}
 						gCopy := g.DeepCopy()
 						gCopy.Status.Listeners = val.Status.Listeners
+
+						for index := range gCopy.Status.Conditions {
```
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:906]: this loop needs to be removed as well

File: internal/provider/kubernetes/controller.go [hunk 5]
```diff
@@ -922,6 +939,12 @@ func (r *gatewayAPIReconciler) subscribeAndUpdateStatus(ctx context.Context) {
 						}
 						hCopy := h.DeepCopy()
 						hCopy.Status.Parents = val.Status.Parents
+
+						for parentIndex, parent := range hCopy.Status.Parents {
```
Comment (arkodg) [inline internal/provider/kubernetes/controller.go:943]: same as above

**Comment by arkodg** [comment]
@Xunzhuo you will also need to edit this code to
```
func conditionChanged(a, b metav1.Condition) bool {
	return a.Status != b.Status || a.Reason != b.Reason || a.Message != b.Message || || a.ObservedGeneration != b.ObservedGeneration
}
```

**Comment by arkodg** [comment]
@Xunzhuo you will also need to rm `ObservedGeneration` from here https://github.com/envoyproxy/gateway/blob/97032bee2db5ec73027b172e24b03d4dfe98ac3a/internal/status/status.go#L164 since it is now a comparable field

File: internal/provider/kubernetes/controller.go (hunk 3 above)
Comment (Xunzhuo) [reply internal/provider/kubernetes/controller.go]: After removing them, CI fails : (

**Comment by arkodg** [comment]
hey @Xunzhuo looks like there was an issue where the gateway API status updater code path would try and update `Status.Listeners` without updating the ObservedGeneraton with `Status.Conditions`  causing the test to fail. Ive updated the logic here https://github.com/envoyproxy/gateway/pull/874

**Pull request closed without merging**